pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -213,14 +216,16 @@ func newQueryTripperware(
 	metrics := newInstrumentMiddlewareMetrics(registerer)
 	queryBlockerMiddleware := newQueryBlockerMiddleware(limits, log, registerer)
 	queryStatsMiddleware := newQueryStatsMiddleware(registerer, engine)
+	federationMiddleware := newFederationMiddleware(federation)
```

Comment:
related to this - can we extend the Authentication HTTP middleware to do this validation instead of making it query-frontend-specific? 

https://github.com/grafana/mimir/blob/6a24a60ac1863595e11db005fbf75d41c97969b7/pkg/api/api.go#L115-L118

Reply:
> how did label values and cardinality requests use to work? Do we need to add this validation there too? I see this middleware is only added on instant and range queries

You're right, we're not applying this logic to label or cardinality requests. I'll see if we can do this in a different place to make sure it applies to other requests.

Reply:
OK, I've got this working locally using `noauth.SetupAuthMiddleware`. Instead of using `middleware.AuthenticateUser` from dskit, we use a new `middleware.Interface` implementation to do the same thing `middleware.AuthenticateUser` was doing and additionally enforce a max number of tenant IDs that may be present.

We have two choices:
* We can keep tenant federation validation as a query-frontend middleware because the read path is the only place we support multiple tenants. This will mean we need to apply the middleware to label and cardinality requests and make sure to add it to any other types of requests we add in the future.
* Or we can modify the basic HTTP middleware we use for all "authenticated" requests. The downside of this is that we might end with confusing combinations of errors messages where this middleware requires fewer than N tenant IDs for request to an endpoint on the write path that only supports a single tenant ID anyway.

WDYT?

Reply:
> we use a new middleware.Interface implementation to do the same thing `middleware.AuthenticateUser` was doing and additionally enforce a max number of tenant IDs that may be present.

you can use `middleware.Merge` to not have to reimplement what `middleware.AuthenticateUser` does

> modify the basic HTTP middleware we use for all "authenticated" requests

this is my preference as it makes this logic opt-out instead of opt-in. The confusing error messages could be a thing, but I would prefer those to an auth bug where we allow some invalid tenants somewhere, but not elsewhere (like if the user can write to the tenant, but only later realize they can't query it after already setting up the ingestion pipeline)

Reply:
I saw you also raised this on slack. Looks like the consensus was to go with the auth middleware.

Reply:
I ended up going with the auth middleware approach for the reasons you mentioned.

I didn't realize `middleware.Merge` was an option and ended up adding the functionality to a new middleware. It's only 3 lines of code so I'm going to leave it unless you feel strongly about it.